File: src/main/java/rars/riscv/lang/lexing/TokensProducer.java
package rars.riscv.lang.lexing;

import javax.swing.text.Segment;

public interface TokensProducer<CollectionType> {
    void addToken(char[] array, int start, int end, RVTokenType tokenType,
                  int startOffset);

    default void addNullToken(char[] array, int segmentPos, int offset) {
        this.addToken(array, segmentPos, segmentPos, RVTokenType.NULL, offset);
    }

    default void addErrorToken(char[] array, int segmentPos, int offset, String notice) {
        this.addToken(array, segmentPos, segmentPos, RVTokenType.ERROR, offset);

    }

    CollectionType getResult();

    CollectionType getEmptyResult();

    CollectionType getTokenList(Segment text, int initialTokenType, int lineOffset, int lineNum);

}

File: src/main/java/rars/riscv/lang/lexing/TokensProducer.kt
package rars.riscv.lang.lexing

import javax.swing.text.Segment

interface TokensProducer<CollectionType : Any?> {
    fun addToken(
        array: CharArray, start: Int, end: Int, tokenType: RVTokenType,
        startOffset: Int
    )

    fun addNullToken(array: CharArray, segmentPos: Int, offset: Int) {
        addToken(array, segmentPos, segmentPos, RVTokenType.NULL, offset)
    }

    fun addErrorToken(array: CharArray, segmentPos: Int, offset: Int, notice: String) {
        addToken(array, segmentPos, segmentPos, RVTokenType.ERROR, offset)
    }

    fun getResult(): CollectionType

    fun getEmptyResult(): CollectionType

    fun getTokenList(text: Segment, initialTokenType: Int, lineOffset: Int, lineNum: Int): CollectionType
}
